source/MaterialXGenShader/ShaderTranslator.cpp

```diff
@@ -0,0 +1,207 @@
+//
+// TM & (c) 2020 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXGenShader/ShaderTranslator.h>
+
+#include <MaterialXCore/Util.h>
+
+#include <iostream>
+
+namespace MaterialX
+{
+
+//
+// ShaderTranslator methods
+//
+
+ShaderTranslator::ShaderTranslator(ConstDocumentPtr doc) :
+    _doc(doc)
+{
+    loadShadingTranslations();
+}
+
+void ShaderTranslator::loadShadingTranslations()
+{
+    for (NodeDefPtr node : _doc->getNodeDefs())
+    {
+        if (node->getNodeGroup() == NodeDef::TRANSLATION_NODE_GROUP)
+        {
+            _translationNodes.insert(node->getNodeString());
+
+            // Parsing translation nodes
+            size_t pos = node->getNodeString().find("_to_");
+            string start = node->getNodeString().substr(0, pos);
+            string end = node->getNodeString().substr(pos + 4);
+            std::unordered_map<string, StringSet>::const_iterator it = _shadingTranslations.find(start);
+
+            if (it != _shadingTranslations.end())
+            {
+                _shadingTranslations[start].insert(end);
+            }
+            else
+            {
+                _shadingTranslations[start] = { end };
+            }
+        }
+    }
+}
+
+void ShaderTranslator::connectToTranslationInputs(ShaderRefPtr shaderRef)
+{
+    for (BindInputPtr bindInput : shaderRef->getBindInputs())
+    {
+
+        OutputPtr output = bindInput->getConnectedOutput();
+        if (output)
+        {
+            InputPtr input = _translationNode->addInput(bindInput->getName(), bindInput->getType());
+            input->setConnectedNode(_graph->getNode(output->getNodeName()));
+
+            _graph->removeOutput(output->getName());
+        }
+        else if (bindInput->getValueString() != EMPTY_STRING)
+        { 
+            InputPtr input = _translationNode->addInput(bindInput->getName(), bindInput->getType());
+            input->setValueString(bindInput->getValueString());
+        }
+        else
+        {
+            std::cerr << "No associated output with " << bindInput->getName() << std::endl;
+        }
+
+        shaderRef->removeBindInput(bindInput->getName());
+    }
+}
+
+void ShaderTranslator::insertOutputNgUpstreamElems(OutputPtr translatedOutput, OutputPtr graphOutput)
+{
+    ElementPtr top = graphOutput->asA<Element>();
+    vector<ElementPtr> upstreamElements;
+    while (top)
+    {
+        upstreamElements.push_back(top);
+        top = top->getUpstreamElement();
+    }
+
+    // rebuilding upstream node dependencies from translation output to input
+    for (ElementPtr upstreamElem : upstreamElements)
+    {
+        if (upstreamElem == graphOutput->asA<Element>())
+        {
+            // connecting output to the upstream node (that will be copied over)
+            translatedOutput->setNodeName(upstreamElem->asA<Output>()->getNodeName());
```

Comment:
Perhaps this logic would be clearer using the built-in Output::setConnectedNode method?